src/bulkUtils.ts

```diff
@@ -174,3 +182,125 @@ export const remainingTime =
   (now: number) =>
   (endWaitTime?: number): number =>
     Math.max((endWaitTime ?? now) - now, 0);
+
+export enum ColumnDelimiter {
+  BACKQUOTE = '`',
+  CARET = '^',
+  COMMA = ',',
+  PIPE = '|',
+  SEMICOLON = ';',
+  TAB = '	',
+}
+
+export type ColumnDelimiterKeys = keyof typeof ColumnDelimiter;
+
+async function bulkRequest(conn: Connection, url: string): Promise<{ body: string; headers: HttpResponse['headers'] }> {
+  const httpApi = new HttpApi(conn, {
+    responseType: 'text/plain', // this ensures jsforce doesn't try parsing the body
+  });
+
+  let headers: HttpResponse['headers'] | undefined;
+
+  httpApi.on('response', (response: HttpResponse) => {
+    headers = response.headers;
+  });
+
+  const body = await httpApi.request<string>({
+    url: conn.normalizeUrl(url),
+    method: 'GET',
+  });
+
+  if (!headers) throw new Error('failed to get HTTP headers for bulk query');
+
+  return {
+    body,
+    headers,
+  };
+}
+
+export async function exportRecords(
```

Comment:
we can't use jsforce's `bulk2.query` because it fetches all batches in memory, does 2/3 passes (raw csv from API -> parse -> back to csv) so you get out of memory quickly mid/big queries.

This helper function fetches 1 batch, parses it (only if output=`json`, API returns batches as CSV so we can write it to a file), and drops it.